src/liballoc/collections/btree/map.rs

```diff
@@ -2001,8 +2001,8 @@ where
             }
         }
 
-        let front = Handle::new_edge(min_node, min_edge);
-        let back = Handle::new_edge(max_node, max_edge);
+        let front = unsafe { Handle::new_edge(min_node, min_edge) };
+        let back = unsafe { Handle::new_edge(max_node, max_edge) };
```

Comment:
I put in a rather long comment, and figured the unsafe code exists only because `range_search` goes to a lower level of abstraction than it has to for correctness. But [lifting it higher](https://github.com/ssomers/rust/tree/btree_search_tidying) causes benchmarking mysteries, i.e. it grinds down iteration benchmarks that never invoke `range_search` to begin with (and I'm used to 10% noise in performance, not 250%), and there are no benchmarks of `range_search` itself. So that's definitely for another PR.

Reply:
> commentary on these for why the preconditions are met here (**and elsewhere**)

Do you propose to (eventually) do that in every invocation? While surrounding the calls with unsafe blocks, I found it easy to confirm preconditions from nearby if-statements or asserts, in all cases except two, which includes the one you quoted here. But of course, it's a lot easier when you're aware of the specific precondition to look for.